controllers/User.php

```diff
@@ -8,50 +8,112 @@
  */
 class User extends MY_Controller {
 
-    public function login()
-    {
-        // Redirect logged in users to home.
-        if ($this->rodsuser->isLoggedIn()) {
+    public function callback() {
+	$code 		= $this->input->get('code', TRUE);
+	$tokenUrl 	= $this->config->item('oidc_token_uri');
+	$callbackUrl 	= $this->config->item('oidc_callback_uri');
+	$clientId 	= $this->config->item('oidc_client_id');'';
+	$clientSecret 	= $this->config->item('oidc_client_secret');
+	$email_field	= $this->config->item('oidc_email_field');
+	$CREDS 		= base64_encode("$clientId:$clientSecret");
+
+	$formdata = array(
+		'grant_type' => 'authorization_code', 
+		'code' => $code, 
+		'redirect_uri' => $callbackUrl
+	);
+
+	$options = [
+		'http' => [
+			  'header' => [ "Authorization: Basic $CREDS"
+			      	      , "Content-Type: application/x-www-form-urlencoded"
+				      ]
+			, 'method' => 'POST'
+			, 'content' => http_build_query($formdata)
+			]
+	];
+
+	$context = stream_context_create($options);
+	$result = file_get_contents($tokenUrl, false, $context);
+	$jsonresult = json_decode($result, TRUE);
+
+	$claimElement = explode('.', $jsonresult['id_token'])[1];
+	$claimData = json_decode(base64_decode($claimElement), TRUE);
+
+        $this->session->unset_userdata('username');
```

Comment:
Whitespace

Reply:
Redid the entire file, replacing tabs by spaces and realigned everything again.